File: 1510-find-lucky-integer-in-an-array/find-lucky-integer-in-an-array.cpp
class Solution {
public:
    /*
        Function    : findLuckyNumber1
        Description : Function used to find the lucky number based on the given problem statement.
        Author      : Pravin Mhaske
        Date        : 05/06/2025
    */
    int findLuckyNumber1(const std::vector<int>& nums) {
        int frequency[501] = {0};

        // Update the frequency of the each number
        for(int num: nums) {
            ++frequency[num];
        }

        // Now check the number whose frequency is equal it itself and it is maximum.
        for( int n = nums.size(); n>0; --n) {
            if(frequency[n] == n) {
                return n;
            }
        }

        return -1;
    }

    /*
        Analysis:
        Time Complexity  : O(n)
        Space Complexity : O(n)
    */

    int findLucky(vector<int>& arr) {
        return findLuckyNumber1(arr);
    }
};

File: 1510-find-lucky-integer-in-an-array/find-lucky-integer-in-an-array.java
class Solution {
    // Function used find the lucky integer
    private int findLuckyNumber(int[] nums) {
        int luckyNumber = -1;
        HashMap<Integer, Integer> frequency = new HashMap<Integer, Integer>();

        // Update the frequecy of the each number
        for (int num: nums) {
            frequency.put(num, frequency.getOrDefault(num, 0) + 1);
        }

        // CHeck the number that could be lucky number
        for(Integer key : frequency.keySet()) {
            if (key == frequency.get(key)) {
                luckyNumber = key;
            }
        }

        return luckyNumber;        
    }

    /*
        Analysis:
        Time Complexity  : O(n)
        Space Complexity : O(n)
    */

    public int findLucky(int[] arr) {
        return findLuckyNumber(arr);
    }
}